ddtrace/tracer/writer.go

```diff
@@ -89,22 +89,38 @@ func (h *agentTraceWriter) flush() {
 	h.payload = newPayload()
 	go func(p *payload) {
 		defer func(start time.Time) {
+			// Once the payload has been read, clear the buffer for garbage collection to avoid
+			// a memory leak when references to this object may still be kept by faulty transport
+			// implementations or the standard library. See dd-trace-go#976
+			p.buf = bytes.Buffer{}
+			p.reader = nil
+
 			<-h.climit
 			h.wg.Done()
 			h.statsd.Timing("datadog.tracer.flush_duration", time.Since(start), nil, 1)
 		}(time.Now())
-		size, count := p.size(), p.itemCount()
-		log.Debug("Sending payload: size: %d traces: %d\n", size, count)
-		rc, err := h.config.transport.send(p)
-		if err != nil {
-			h.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
-			log.Error("lost %d traces: %v", count, err)
-		} else {
-			h.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
-			h.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
-			if err := h.prioritySampling.readRatesJSON(rc); err != nil {
-				h.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+
+		for attempt := 0; attempt <= h.config.sendRetries; attempt++ {
+			size, count := p.size(), p.itemCount()
+			log.Debug("Sending payload: size: %d traces: %d\n", size, count)
+			rc, err := h.config.transport.send(p)
+			if err == nil {
+				log.Debug("sent traces after %d attempts", attempt+1)
+				h.statsd.Count("datadog.tracer.flush_bytes", int64(size), nil, 1)
+				h.statsd.Count("datadog.tracer.flush_traces", int64(count), nil, 1)
+				if err := h.prioritySampling.readRatesJSON(rc); err != nil {
+					h.statsd.Incr("datadog.tracer.decode_error", nil, 1)
+				}
+				return
+			}
+			if attempt == h.config.sendRetries {
+				h.statsd.Count("datadog.tracer.traces_dropped", int64(count), []string{"reason:send_failed"}, 1)
+				log.Error("lost %d traces: %v", count, err)
```

Comment:
These 2 lines can be moved after the `for` loop, and we can get rid of this `if` statement.